source/MRViewer/MRViewer.cpp

```diff
@@ -1459,6 +1460,63 @@ bool Viewer::draw_( bool force )
     return ( window && swapped );
 }
 
+void Viewer::drawUiRenderObjects_()
+{
+    // Currently, a part of the contract of `IRenderObject::renderUi()` is that at most rendering task is in flight at any given time.
+    // That's why each viewport is being drawn separately.
+
+    for ( Viewport& viewport : getViewerInstance().viewport_list )
+    {
+        UiRenderParams renderParams;
+        renderParams.scale = menuPlugin_->menu_scaling();
+        renderParams.viewport = &viewport;
+
+        auto rawViewportRect = viewport.getViewportRect();
+        renderParams.viewportCornerA = rawViewportRect.min;
+        renderParams.viewportCornerB = rawViewportRect.max;
+        renderParams.viewportCornerA.y = ImGui::GetIO().DisplaySize.y - renderParams.viewportCornerA.y;
+        renderParams.viewportCornerB.y = ImGui::GetIO().DisplaySize.y - renderParams.viewportCornerB.y;
+        std::swap( renderParams.viewportCornerA.y, renderParams.viewportCornerB.y );
+
+        // Set up clipping as a courtesy.
+        ImDrawList* drawLists[] = { ImGui::GetBackgroundDrawList(), ImGui::GetForegroundDrawList() };
+        for ( ImDrawList* list : drawLists )
+            list->PushClipRect( renderParams.viewportCornerA, renderParams.viewportCornerB );
+        MR_FINALLY{
+            for ( ImDrawList* list : drawLists )
+                list->PopClipRect();
+        };
+
+        IRenderObject::UiTaskList tasks;
+        tasks.reserve( 50 );
```

Comment:
all imgui logic should be inside render object implementation

Reply:
Moved to ImGuiMenu.